web/src/components/text-input/generic.tsx

```diff
@@ -0,0 +1,137 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import React, { useState } from 'react'
+import {
+  AriaTextFieldOptions,
+  TextFieldAria,
+  useTextField,
+} from '@react-aria/textfield'
+import clsx from 'clsx'
+
+import { Text } from '../typography'
+import useStyles from './styles'
+import { useFocusStyle } from '../../styles/global'
+import { Flex } from '../layout/flex'
+import { Icon } from '../icon'
+
+export interface TextInputProps extends AriaTextFieldOptions {
+  id?: string
+  isInvalid?: boolean
+  placeholder?: string
+  description?: string
+  validationHelp?: string
+  renderLeft?: () => JSX.Element
+  renderRight?: (props: { onClear: () => void; dirty?: boolean }) => JSX.Element
+}
+
+export interface WithoutLabel extends TextInputProps {
+  'aria-label': string
+}
+
+export interface WithLabel extends TextInputProps {
+  label: string
+}
+
+export type InputProps = WithoutLabel | WithLabel
+
+export const GenericField: React.FC<
+  InputProps & { component: 'input' | 'textarea' }
+> = props => {
+  const {
+    label,
+    isInvalid,
+    description,
+    validationHelp,
+    maxLength,
+    component,
+    isRequired,
+    renderLeft,
+    renderRight,
+  } = props
+  const ref = React.useRef<HTMLInputElement & HTMLTextAreaElement>()
+  const [allowedChars, setAllowedChars] = useState(
+    maxLength - (props.value?.length ?? 0),
+  )
+  const [dirty, setDirty] = useState(false)
+
+  const { labelProps, inputProps } = useTextField(
+    {
+      ...props,
+      maxLength: undefined,
+      onChange: p => {
+        setDirty(p.length > 0)
+        props.onChange?.(p)
+        setAllowedChars(maxLength - p.length)
+      },
+    },
+    ref,
+  )
+  const invalid = isInvalid || allowedChars < 0
+  const customProps = { ...props, isInvalid: invalid }
+  const { onFocus } = useFocusStyle(customProps)
+  const classes = useStyles(customProps)
+
+  const Component = component
+
+  return (
+    <Flex direction="column" gap={1}>
+      {label && (
+        <label className={classes.label} {...labelProps}>
+          {label}
+          {isRequired !== undefined && (
+            <Text
+              color="secondary"
+              size="body-small"
+              as="span"
+              className={classes.suffix}
+            >
+              ({isRequired ? 'required' : 'optional'})
+            </Text>
+          )}
+        </label>
+      )}
+      {validationHelp && (
+        <Flex gap={1}>
+          <Icon name="alert" color="danger" size="small" />
+          <Text as="span" size="body-small" color="danger">
+            {validationHelp}
+          </Text>
+        </Flex>
+      )}
+      <div style={{ position: 'relative' }}>
+        {renderLeft?.()}
+
+        <Component
+          className={clsx(classes.textField, onFocus)}
+          {...inputProps}
+          {...((invalid || allowedChars < 0) && { 'aria-invalid': true })}
```

Comment:
But invalid already contain check for `allowedChars < 0`. Shouldn't this check be removed here? Or use `isInvalid` here and below and remove `invalid`.

Reply:
good catch thanks!